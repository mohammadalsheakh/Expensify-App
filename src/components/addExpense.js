import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./expenseForm";
import { addExpense } from "../actions/expenses";
import { useNavigate } from "react-router-dom";


// connect function two types :
//state to props and dispatch to props , here we are not using any but we need to pass the dispatch to props as 
//the second argument becouse we have a proplem in testing with this code so :


// const AddExpense = (props) => {
//     const navigate = useNavigate()
//     const GetTheExpense = (expense) =>{
//         console.log(expense);
//         props.dispatch(addExpense(expense))
//         navigate(`/`)
        
//     }
//     return(
//         <div>
//             Add expense here gg
//             < ExpenseForm  
//             getTheExpense={GetTheExpense}
//             />
//         </div>
//     )
// }
// export default connect()(AddExpense)




// here we have the new code :
export const AddExpense = (props) => {
    const navigate = useNavigate()
    const GetTheExpense = (expense) => {
        props.onDispatch(expense)
        navigate(`/`)

    }
    return (
        <div>
            Add Expense 
            < ExpenseForm
                getTheExpense={GetTheExpense}
            />
        </div>
    )
}

const mapDispatchToProps = (dispatch) => ({ // here we used the dispatch insted of the state 
    onDispatch: expense => dispatch(addExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpense)