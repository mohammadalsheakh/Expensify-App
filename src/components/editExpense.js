import React from "react";
import { Link, Outlet } from "react-router-dom";

const EditExpense = () => (
    <div>
        edit expense here
        <Outlet />
    </div>
)
export default EditExpense