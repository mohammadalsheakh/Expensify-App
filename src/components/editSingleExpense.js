import React from 'react'
import { useParams } from 'react-router-dom'
import { useNavigate } from 'react-router-dom'
import { connect } from 'react-redux'
import ExpenseForm from './expenseForm'
import {editExpense, removeExpense} from '../actions/expenses'



function EditSingleExpense(props) {
    let navigate = useNavigate()
    let params = useParams()
    
    const getNeeded = () =>{
        return (props.expenses.find((expense) => {
            return expense.id === params.id
        }))
        
    }
    let theExpense = getNeeded()

    const GetTheExpense = (expense) => {
        props.dispatch(editExpense(theExpense.id, expense ))
        navigate(`/`)
    }


    return (
        <div>
            <ExpenseForm 
                expense={theExpense}
                getTheExpense={GetTheExpense}
            />
            <button onClick={() => {
                props.dispatch(removeExpense({ id: theExpense.id }))
                navigate(`/`)
            }} >remove</button>
        </div>
        
    )
}

const mapStateToProps = (state) =>{
    
    return{
        expenses: state.expenses}
    }
    


export default connect(mapStateToProps)(EditSingleExpense)