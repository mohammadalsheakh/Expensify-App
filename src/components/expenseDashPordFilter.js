import React from 'react'
import { connect } from "react-redux";
import { setTextFilter} from '../actions/filters'
import { DateRangePicker } from 'react-dates';
import { sortByDate , sortByAmount, setEndDate, setStartDate} from '../actions/filters'
import ExpenseListItem from './expenseListItem';

export class ExpenseDashPordFilter extends React.Component{

    state = {
        clenderFoucused : null
    }

    onDatesChange = ({ startDate, endDate}) =>{
        this.props.setEndDate(endDate)
        this.props.setStartDate(startDate)
    }
    onFocusChange = (clenderFoucused) =>{
        this.setState(() => ({clenderFoucused}))
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        if (e.target.value === `date`) {
            this.props.sortByDate()
        }
        else if (e.target.value === `amount`) {
            this.props.sortByAmount()
        }
    }
    render(){
        return(
            <div>
                <select value={this.props.filterValue.sortBy} onChange={this.onSortChange} >
                    <option value="date" key="date">Date</option>
                    <option value="amount" key="amount">Amount</option>
                </select>
                <input type="text" value={this.props.filterValue.text} onChange={this.onTextChange} />
                <DateRangePicker 
                startDate={this.props.filterValue.startDate}
                startDateId="My Start Date"
                endDate={this.props.filterValue.endDate}
                endDateId="My End Date"
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.clenderFoucused}
                onFocusChange={this.onFocusChange}
                horizontalMargin={0}
                numberOfMonths={1}
                isOutsideRange={() => false}
                screenReaderInputMessage="pick range"
                showClearDates={true}

                />
            </div>

        )
    }
}

const connectMe = (state) =>{
    return{
        filterValue : state.filters
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter : text => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate : () => dispatch(sortByDate()),
    setStartDate : (startDate) => dispatch(setStartDate(startDate)),
    setEndDate : (endDate) => dispatch(setEndDate(endDate))
    
})
export default connect(connectMe,mapDispatchToProps)(ExpenseDashPordFilter)