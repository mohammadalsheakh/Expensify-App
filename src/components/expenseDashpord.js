import React from "react";
import { connect } from "react-redux";
// import { useNavigate } from "react-router-dom";
import ExpensesList from "./expenselist";

const EditDashbord = () => {

    // let navigate = useNavigate()
    //     < button onClick = {() => { navigate("/help") }}> click</button >
    return (
        <div>
            <ExpensesList />
        </div>
    )
}


export default EditDashbord