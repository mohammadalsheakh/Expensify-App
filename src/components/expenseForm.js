import React, { Component } from 'react'
import moment from 'moment'
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';




export default class ExpenseForm extends Component {

    constructor(props){
        super(props)
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount: ``,
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ``
        }

    }
    
    
    onDescriptionChange = (e) =>{
        let description = e.target.value
        this.setState(() => ({ description}))
    }

    onAmountChagne = (e) =>{
        const solve = e.target.value
        let note = ``
        if (solve.includes(`.`)) {
            note = solve.slice(0, solve.indexOf(`.`) + 2)
        }
        else note = solve.slice(0,solve.length)
        this.setState(() => ({amount : note}))
    }

    onNoteChange = (e) =>{
        let noteChange = e.target.value
        this.setState(() => ({note : noteChange}))
    }

    onDateChange = (createdAt) =>{ // here we have created at representes the variable that contaian the current date 
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    }

    onFocusChange = ({focused}) => {
        this.setState({ calendarFocused: focused})
    }

    onSubmit = (e) =>{
        e.preventDefault()
        if (!this.state.description  || !this.state.amount) {
            this.setState({error : `Pleas provide amount and description`})
        }
        else {
            this.setState({ error: `` })
            this.props.getTheExpense({
                description : this.state.description,
                amount : Number.parseFloat(this.state.amount),
                createdAt : this.state.createdAt.valueOf(),
                note : this.state.note
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type="text"   placeholder='Description' autoFocus value={this.state.description} onChange={this.onDescriptionChange} />
                    <input type="number"  placeholder='amount' value={this.state.amount} onChange={this.onAmountChagne} />
                    <SingleDatePicker 
                        
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        screenReaderInputMessage={`chose date`}
                        horizontalMargin={0}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea placeholder='You can add a note here' value={this.state.note} onChange={this.onNoteChange}></textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}

