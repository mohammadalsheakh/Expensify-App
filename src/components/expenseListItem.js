import React from 'react'
import { Link, Outlet } from "react-router-dom";


function ExpenseListItem({ id ,description , amount , createdAt}) {
    return (
        <div>
        <Link to={`edit/${id}`}>
                <h3 style={{ color :"red"}}>{description}</h3>
        </Link>
            
            <p>{amount} - {createdAt}</p>
            
        </div>
    )
}

export default (ExpenseListItem)