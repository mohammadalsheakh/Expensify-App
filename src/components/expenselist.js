import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./expenseListItem";
import getVisibleExpenses from "../selections/expenses";
import ExpenseDashPordFilter from "./expenseDashPordFilter";

export const ExpensesList = (props) => (
    <div>
        <ExpenseDashPordFilter />
        
        {
            props.expense.length === 0 ? (<p>There is no expenses</p>)  : 
                (props.expense.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense} />
                }))
        }

    </div>
)

//this usage of the connect api is an application for higher order component 
const mapStateToProps = (state) => {
    return {
        expense: getVisibleExpenses(state.expenses, state.filters)
    }
}
const ConnectedEditList = connect(mapStateToProps)(ExpensesList)



// so connect api will return an object to the child componenet 

export default ConnectedEditList