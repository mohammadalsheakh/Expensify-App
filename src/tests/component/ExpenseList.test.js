import React from "react";
import { shallow } from "enzyme";
import { ExpensesList } from "../../components/expenselist";
import myExpenses from "../fixture/myexpenses";


test('should render the exoense list with expenses', () => {
    const wraper = shallow(< ExpensesList expense={myExpenses} />)
    expect(wraper).toMatchSnapshot()
})

test('should render when there is no expenses', () => {
    const wraper = shallow(< ExpensesList expense={[]} />)
    expect(wraper).toMatchSnapshot()
})

