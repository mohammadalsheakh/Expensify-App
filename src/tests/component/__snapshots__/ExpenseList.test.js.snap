// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the exoense list with expenses 1`] = `
<div>
  <Connect(ExpenseDashPordFilter) />
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id={1}
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={109}
    createdAt={-345600000}
    description="Rent"
    id={2}
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={45}
    createdAt={345600000}
    description="credit card"
    id={3}
    key="3"
    note=""
  />
</div>
`;

exports[`should render when there is no expenses 1`] = `
<div>
  <Connect(ExpenseDashPordFilter) />
  <p>
    There is no expenses
  </p>
</div>
`;
